Hvg 100 add details section on web on boarding

virrios opened the pull request:
Added Details views for Denmark and Norway
*The design is not updated, as well as I am just using the fields that are Danish/Norwegian specifik.
A suggestion for restructure the Details view since it had to be split up, otherwise it would have been too much code in one place.  
*Using a shared styledComponents file between the different details. Open for naming suggestions.

There are some other fields that can be included in the update Details however I felt like this was a big pr in it self and think should be approved before additional change are included especially if the Design is not finished.

Tested To update SE/DK/NO details after debugger and using the embark flow. (Not tested every permutation...)

Potential Code reuse can be improved however, I felt like it would be nice to split stuff into the different markets, so you can local changes without having to alter some general code that needs to handle all the different cases.

Commit:
Refactored some code and folder structure to make it not as cluttured when adding
Details sections for more markets.

Commit:
Added Norway and Danish attributes to Quote & QuoteBundle Queries

  Obs type uses alias since type  have different potential types which causes a
  graphql error if you use type for both swe/no/dk.
  A mapping is required later.

  Generated the new graphql file as well.

Commit:
Added types to handle Danish and Norwegian details

Commit:
Updated types for Details and move out to separate file to so it can used by other market details

Commit:
Added Norwegian details

Commit:
Added Danish details

Commit:
Added functionality in utils to handle Norway and Denmark
 Updated mask to handle that zipcodes for Se is handled different compared to No and Dk

Commit:
Show details link and modal for denmark and norway aswell

Commit:
Create Details component that wraps the market

Commit:
Updated Translations

General update aswell as for keystring:
  *  DETAILS_MODULE_TABLE_YOUTH_CELL_LABEL
  * DETAILS_MODULE_TABLE_STUDENT_CELL_LABEL

Commit:
Updated isYouth and isStudent with correct localize keys

palmenhq commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/components/DetailInput.tsx:
```diff
@@ -3,12 +3,18 @@ import React from 'react'
+    | 'apartment'
+    | 'house'
+    | 'norwegianHomeContents'
+    | 'norwegianTravel'
+    | 'norwegianTravel'
```
travel travel?

virrios replied:
🙈

palmenhq reviewed (commented):
Looks awesome! A few questions tho

virrios commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/components/index.tsx:
```diff
@@ -0,0 +1,3 @@
+export * from './styledComponents'
```
Yes, I was going back and forth between those names! The thing about styles is that it might be confused with only styling objects, but since you only use emotions that is maybe not the case. I will change!

gustaveen replied:
Should we really use index files for exports? I would argue for just using named exports and get rid of extra files 🙂

gustaveen replied:
Also, I agree with that we should use`.styles` but also some more context to the filename. Like in this case it's styling for the Details components so what do you think about `Details.styles.tsx` ?

virrios replied:
I like the Details.styles.tsx suggestions because it gets easier to navigate the project, if  you search the project for Details. Then you see the Details.styles file as well.

virrios commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/components/styledComponents.tsx:
```diff
@@ -0,0 +1,66 @@
+import styled from '@emotion/styled'
+import { colorsV2 } from '@hedviginsurance/brand'
```
I only moved the styles. Should This be part of the this pr, I can do it but just so I know? (same for the styling comments below)

palmenhq replied:
Ah yes fair point.. Maybe you could skip it for now, it's mostly different shades of gray so itt can't be that critical haha 😄

virrios commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/index.tsx:
```diff
@@ -0,0 +1,43 @@
+import { SwedishDetails } from './SwedishDetails'
+import { Details } from './types'
+
+export default ({
+  fieldSchema,
+  formikProps,
+  offerQuote,
+  textKeys,
+}: Details) => {
```
I noticed that you only used that. However if you have the type in a different file you cant have the same name for the component. I can change it and maybe name Details Type to.....DetailsProps, or any other suggestions?:)

palmenhq replied:
Right! i think `DetailsProps` make more sense, optionally if you use an import alias in this file :)

styrelius replied:
Still a bit confusing if you use the same name for types and components imo :)

virrios commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/utils.ts:
```diff
@@ -41,33 +83,20 @@ export const isHouseFieldSchema = (
-      type: inputTypes.string,
-      validation: Yup.string().matches(/^[0-9]{3}[0-9]{2}$/),
-    },
+const getSwedishSchema = (base: any, offerQuote: OfferQuote) => {
```
I missed typing that! No point in having any!

virrios commented on src/client/components/inputs/index.tsx:
```diff
@@ -21,6 +21,11 @@ export const masks: Record<string, Mask> = {
     mask: '999 99',
     sanitize: (value) => value.replace(/\s+/, ''),
   },
+  zipCodeNoDk: {
+    name: 'ZipCode',
+    mask: '9999',
+    sanitize: (value) => value.replace(/\s+/, ''),
+  },
```
I think fiveDigitZipCode and fourDigitZipCode, Think it would be easier to spot the errors then?

virrios commented on src/client/pages/OfferNew/Introduction/Sidebar/index.tsx:
```diff
@@ -302,14 +300,12 @@ export const Sidebar = React.forwardRef<HTMLDivElement, Props>(
-                />
-              )}
+              <DetailsModal
+                offerQuote={offerData.quotes[0]}
```
I am not sure about how it should work with bundles. I think that this might be a different pr when we don't have the complete design?  But in any case I would need a debrief 🤓

palmenhq replied:
Sounds wise, but lets disable this for norway in that case, just until we handle that :) so we don't risk getting one quote with 1 co-insured and another with 2, for example :)

gustaveen replied:
Ah good catch, didn't think about how to treat bundles either. We might need to do so visual clarification as well so the user knows what they are editing

virrios commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/utils.ts:
```diff
@@ -340,14 +516,16 @@ export const getInitialSwedishHouseValues = (
   },
 })
-
+type NorwegianHomeContentsDetailsAlias = NorwegianHomeContentsDetails & {
```
The purpose of the alias is that you have to use alias for type  on NorwegianHomeContentsDetails, DanishHomeContentsDetails. since "type" can be different things. (If you try to generate the graphql it will throw an error) So here the mapping from danishHomeType-> type occurs. 
Maybe I can try to explain it a bit better. because There might be a better solution/naming

palmenhq replied:
Right! Yeah lets have a chat maybe :)

gustaveen commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/components/styledComponents.tsx:
```diff
@@ -0,0 +1,66 @@
+  padding: 0.375rem 1rem;
+  font-size: 0.875rem;
+  color: ${colorsV2.gray};
+  border-radius: 24px;
```
True, for buttons at least. 👍 This is the outlined button right? I thought we could use the button component we already have but I realise that we don't have an outlined one in web-onboarding. So, I think buttons might be the first component for me to add in the brand repo so we can use that instead 😄

gustaveen commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/components/styledComponents.tsx:
```diff
@@ -0,0 +1,66 @@
+  align-items: center;
+  min-height: 2rem;
+
+  @media (max-width: 800px) {
```
This might not be the time to do it if this is old styling, but if we're writing new styles I think we should write mobile-first (min-width) media queries. :)

virrios replied:
Yes it is old styling, but agree small->bigger screen when applying styles!

Commit:
Updated Type checks for FieldSchema in DetailsModal utils

Commit:
Updated Type for intialValues in DetailsModal utils to handle grapql type alias

  Updated Type to use QuoteBundle (requires typescript >=3.9)
  (Had to cast offerQutoe.quoteDetails first (linter suggested unknown but I thought
  QueryQuoteDetails) were a bit clearer.

Commit:
Removed Hacky includes(<Market>) and used isDanishQuote, isSwedishQuote and isNorwegianQuote instead.

   Added helpfunctions isDanishQuote, isSwedishQuote and isNoreqegianQuote
   added isDanish and refactored isSwedish/isNorwegian
   Alo included isDanish in the test suite.

Commit:
Updated mask names for zipCodes.

    * zipCode->fiveDigitZipCode
    * zipCodeNoDk -> fourDigitZipCode

Commit:
Added type to base and removed unnuced imports

Commit:
Fixed Types in Detail and named export

   * Details as named export
   * Details type -> DetailsProps
   * textKeys: any -> textKets: TextKeyMap

Commit:
Removed duplicate norwegianTravel from DetailInputProps

Commit:
Renamed styles.tsx (in Details/components) to Details.styles.tsx

Commit:
Removed index file and used named imports instead

Commit:
Hide Details for Norway since we have not solved the potential double quote for boundles

palmenhq commented on src/client/pages/OfferNew/utils.test.ts:
```diff
@@ -1,10 +1,11 @@
   it('validates if quote is norwegian', () => {
     expect(isNorwegian(noComboYouth)).toBe(true)
     expect(isSwedish(noComboYouth)).toBe(false)
+    expect(isDanish(noComboYouth)).toBe(false)
```
ooh nice. Actually, shall we add a test for danish? quotes?

Commit:
Added test for isDanish and mocked OfferData for DanishHomeContentsDetails

Commit:
Updated colors from colorsv2 to colorsv3

   * gray -> gray500
   * darkgray -> gray900

Commit:
Moved media queries for mobile to the top of the styling

Commit:
Merge branch 'master' into HVG-100-add-details-section-on-web-on-boarding

palmenhq commented on src/client/utils/testData/offerDataMock.ts:
```diff
@@ -217,3 +218,58 @@ export const noCombo: OfferData = {
+    lastName: 'Madsson',
+    email: 'mail@test.com',
+    ssn: '15126533427',
+    birthDate: '2965-12-15',
```
![](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fmedia.giphy.com%2Fmedia%2Fl3nFh9xbJtZYc6UhO%2Fgiphy.gif&f=1&nofb=1)

Commit:
Merge branch 'master' into HVG-100-add-details-section-on-web-on-boarding

styrelius reviewed (commented):
Wow, this is a LOT of work 💪 Had a few questions though!

palmenhq commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/SwedishDetails.tsx:
```diff
@@ -0,0 +1,236 @@
+} from './components/Details.styles'
+import { DetailsProps } from './types'
+
+export const SwedishDetails = ({
+  fieldSchema,
+  formikProps,
+  offerQuote,
+  textKeys,
+}: DetailsProps) => (
+  <>
```
And i guess same here, i'd prefer `React.FC` :)

palmenhq commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/utils.ts:
```diff
@@ -390,7 +608,9 @@ export const getInitialInputValues = (offerQuote: OfferQuote) =>
         getInitialNorwegianHomeContentValues(
           offerQuote.id,
-          offerQuote.quoteDetails as NorwegianHomeContentsDetails,
+          (offerQuote.quoteDetails as QueryQuoteDetails) as QueryQuoteDetails & {
+            __typename: 'NorwegianHomeContentsDetails'
```
This double type cast looks weird to me but maybe it's required? Idk

virrios replied:
I have to cast it, but you can cast it to unknown/any but I thought it might be better to be more exact.

palmenhq commented on src/client/utils/testData/offerDataMock.ts:
```diff
@@ -217,3 +218,58 @@ export const noCombo: OfferData = {
+    freeUntil: null,
+    monthlyDiscount: {
+      amount: '49.00',
+      currency: 'DEK',
```
`DKK`* :D

palmenhq commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/index.tsx:
```diff
@@ -0,0 +1,43 @@
+  fieldSchema,
+  formikProps,
+  offerQuote,
+  textKeys,
```
Hm, i didn't see this before but any partucular reason for passing the text keys as a property? If not i think we should try and use the `useTextKeys` hook inside the components that need them instead. Feels a bit messy to pass it as a prop in some places and use the hook in other.

palmenhq commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/components/Details.styles.tsx:
```diff
@@ -0,0 +1,63 @@
+import styled from '@emotion/styled'
```
In general i agree here! But - since these styles are specific enough for the details form, but they cannot be placed in the `index.tsx` file (since the dependencies would become circular if the files imported from `index.tsx` depended on `index.tsx`) so i see the reasoning for placing them in a special file for this use case. I like this naming more than the previous, but maybe we could work on it more 😅  maybe `formStyles`  or something? Imo it's encapuslated as "the component of Details" and not hard to make out that these styles belong with the sibling files. It'd be en annan femma if it was cross reference to a different component or what do you think?

palmenhq commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/index.tsx:
```diff
@@ -0,0 +1,43 @@
+  formikProps,
+  offerQuote,
+  textKeys,
+}: DetailsProps) => {
```
also, let's use `React.FC`:)

palmenhq reviewed (commented):
gr8!

styrelius commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/NorwegianDetails.tsx:
```diff
@@ -0,0 +1,135 @@
+} from './components/Details.styles'
+import { DetailsProps } from './types'
+
+export const NorwegianDetails = ({
+  fieldSchema,
+  formikProps,
+  offerQuote,
+  textKeys,
+}: DetailsProps) => (
```
I actually disagree about it being more clear - I think it's harder to read and looks... messier. And also my understanding is that both ways are nowadays equally correct, for example according to [this cheatsheet](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components) that is linked to from the TypeScript docs.

But with that said - since we do it the `export const MyComponent: React.FC<SomeProps>` way in all other files I agree about changing it here! I think it's way more important that we are consistent 😇

palmenhq replied:
Interesting and good read! I guess the easiness-to-read and messiness level of it is about habit because I do feel the same about writing the prop type as the prop argument haha 😅

Commit:
Changed type for detail components to  React.FC<type>

Commit:
Removed textKey Prop from Details and it children

Commit:
Merge branch 'master' into HVG-100-add-details-section-on-web-on-boarding

Commit:
DEK-> DKK in offerMockData

styrelius commented:
OK maybe I'm completely lost (not unlikely 🤕 ) but I can't find the link to the (edit) details modal on the Norwegian offer page. Shouldn't it be here?
<img width="1822" alt="2020-11-19-no-offer-page" src="https://user-images.githubusercontent.com/42962286/99651364-caa07700-2a56-11eb-832b-34c99b6997ee.png">

Commit:
Updated from the old parameter to the new ones when doing edit Quote for Sweden.

    Previously it used the "deprecated" house and apartment,
    but now uses swedishHouse and swedishApartment from the GraphQL Schema
    EditApartmentInput, EditHouseInput and EditSwedishApartmentInput and EditSwedishHouseInput
    have the same attributes, respectively.

Commit:
Merge branch 'master' into HVG-100-add-details-section-on-web-on-boarding

virrios commented:
Based on the discussion regarding the potential risk of double quote in Norway I interpreted it as to disable the view in Norway for now?

styrelius commented:
Unfortunately it seems like the changes aren't saved properly when you edit a Danish quote? It works fine for Swedish ones but when I change things on /dk I still get the same ol' details. 🤔 When looking at the Network tab it seems like we get the old, not updated, quote back from the query that we make right after the `EditQuote` mutation. 

![2020-11-19-dk-edit-details-no-update](https://user-images.githubusercontent.com/42962286/99670145-5d014480-2a70-11eb-9644-c784f61bdefe.gif)

![2020-11-19-dk-update-quote-student](https://user-images.githubusercontent.com/42962286/99670167-64285280-2a70-11eb-90fd-d14fbe6d0de8.gif)

Commit:
Merge branch 'master' into HVG-100-add-details-section-on-web-on-boarding

palmenhq commented on src/client/pages/OfferNew/Introduction/Sidebar/DetailsModal/Details/SwedishDetails.tsx:
```diff
@@ -0,0 +1,245 @@
+  InputGroupDeleteButton,
+  InputGroupRow,
+} from 'components/inputs'
+import { ExtraBuildingInput, ExtraBuildingType } from 'data/graphql'
+import { EditQuoteInput } from 'data/graphql'
```
i just saw this, how does it pass eslint? We don't have a rule for multiple imports from the same file?  🤨

palmenhq reviewed (approved):
looks great imo!

<img src="https://external-content.duckduckgo.com/iu/?u=http%3A%2F%2Freactiongif.org%2Fwp-content%2Fuploads%2FGIF%2F2014%2F10%2FAl-Bundy-Approval-Approves-Do-like-Likes-Thumbs-up-I-Like-Good-Job-You-go-boy-You-go-girl-Like-Upvote-GIF.gif&f=1&nofb=1" width=150 />
